runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/source/FlinkSourceSplitEnumerator.java

```diff
@@ -0,0 +1,181 @@
+              i++;
+            }
+            return flinkSourceSplitsList;
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        },
+        (sourceSplits, error) -> {
+          if (error != null) {
+            throw new RuntimeException("Failed to start source enumerator.", error);
+          } else {
+            pendingSplits.putAll(sourceSplits);
+            splitsInitialized = true;
+            sendPendingSplitsToSourceReaders();
+          }
+        });
+  }
+
+  @Override
+  public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+    // Not used.
+  }
+
+  @Override
+  public void addSplitsBack(List<FlinkSourceSplit<T>> splits, int subtaskId) {
+    LOG.info("Adding splits {} back from subtask {}", splits, subtaskId);
+    List<FlinkSourceSplit<T>> splitsForSubtask =
+        pendingSplits.computeIfAbsent(subtaskId, ignored -> new ArrayList<>());
+    splitsForSubtask.addAll(splits);
+  }
+
+  @Override
+  public void addReader(int subtaskId) {
+    List<FlinkSourceSplit<T>> splitsForSubtask = pendingSplits.remove(subtaskId);
+    if (splitsForSubtask != null) {
+      assignSplitsAndLog(splitsForSubtask, subtaskId);
+      pendingSplits.remove(subtaskId);
+    } else {
+      if (splitsInitialized) {
+        LOG.info("There is no split for subtask {}. Signaling no more splits.", subtaskId);
+        context.signalNoMoreSplits(subtaskId);
+      }
+    }
+  }
+
+  @Override
+  public Map<Integer, List<FlinkSourceSplit<T>>> snapshotState(long checkpointId) throws Exception {
+    LOG.info("Taking snapshot for checkpoint {}", checkpointId);
+    return snapshotState();
+  }
+
+  @Override
+  public Map<Integer, List<FlinkSourceSplit<T>>> snapshotState() throws Exception {
+    return pendingSplits;
+  }
+
+  @Override
+  public void close() throws IOException {
+    // NoOp
+  }
+
+  // -------------- Private helper methods ----------------------
+  private List<? extends Source<T>> splitBeamSource() throws Exception {
+    if (beamSource instanceof BoundedSource) {
+      BoundedSource<T> boundedSource = (BoundedSource<T>) beamSource;
+      long desiredSizeBytes = boundedSource.getEstimatedSizeBytes(pipelineOptions) / numSplits;
+      return boundedSource.split(desiredSizeBytes, pipelineOptions);
+    } else if (beamSource instanceof UnboundedSource) {
+      return ((UnboundedSource<T, ?>) beamSource).split(numSplits, pipelineOptions);
+    } else {
+      throw new IllegalStateException("Unknown source type " + beamSource.getClass());
+    }
+  }
+
+  private void sendPendingSplitsToSourceReaders() {
+    Iterator<Map.Entry<Integer, List<FlinkSourceSplit<T>>>> splitIter =
+        pendingSplits.entrySet().iterator();
+    while (splitIter.hasNext()) {
+      Map.Entry<Integer, List<FlinkSourceSplit<T>>> entry = splitIter.next();
+      int readerIndex = entry.getKey();
+      int targetSubtask = readerIndex % context.currentParallelism();
```

Comment:
Not sure if I follow the comment. The key will be in the `pendingSplits` map because that is an "intended assignments" regardless of whether the target reader has actually registered to the `SplitEnumerator` or not. But we can only send the "intended assignments" to a reader if that reader has registered to the `SplitEnumerator`.  That is what the check does.